### PR TITLE
use Guid Codec for UUIDs

### PR DESCRIPTION
**Pull request by prolic**
/cc @enumag

### CODE REVIEW

**Comment by enumag** [comment]
After this `Prooph\EventStoreClient\Util\Uuid` is the only place where ramsey/uuid is used, right? And therefore should be easier to replace. Good work!

**Comment by prolic** [comment]
@enumag Yes, but the main problem here is that we need to use a specific uuid factory to use Guid's to be compatible with the C# implementation, otherwise you'll receive wrong event ids from the http api.

**Comment by enumag** [comment]
I'm not so knowledgeable about UUIDs. What's the difference between UUID4 and GUID?

**Comment by prolic** [comment]
It's a uuid as well, but the order is different, f.e.:

`739c4aaa-3706-4171-a9cc-35ced1a998af` VS `aa4a9c73-0637-7141-a9cc-35ced1a998af`

If you look closely enough, it's really the same numbers, but arranged differently. Guids are used in C#, so I have to force Ramsey Uuid to use Guid as encoding.

**Comment by enumag** [comment]
Oh I see now. You're forcing it with `self::$factory = new UuidFactory(new FeatureSet(true));`. How did you find out this is necessary?

[this](https://en.wikipedia.org/wiki/Universally_unique_identifier#Encoding) seems to be the details about the byte order. We might need that later if we decide to use a different implementation.

File: src/Util/Uuid.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+use Ramsey\Uuid\FeatureSet;
+use Ramsey\Uuid\UuidFactory;
+use Ramsey\Uuid\UuidInterface;
+
+class Uuid
+{
+    /** @var UuidFactory */
+    private static $factory;
+
+    public static function generate(): UuidInterface
```
Comment (enumag) [inline src/Util/Uuid.php]: Shouldn't this and other methods returning UuidInterface be private? This way there are still some ramsey/uuid instances used elsewhere.

Comment (prolic) [reply src/Util/Uuid.php]: Not sure at the moment. I still use it in `EventId::generate()`

Comment (enumag) [reply src/Util/Uuid.php]: Yeah I noticed. What is the reason?

**Comment by prolic** [comment]
How did I find out? I did a reference to another event id on a work project and that event id never matched the one that I could see in the Event Store UI. Of course I didn't see the order difference the first time, so I began debugging and finally found this out. It took me like 2 hours maybe.

File: src/Util/Uuid.php (hunk 1 above)
Comment (prolic) [reply src/Util/Uuid.php]: On the `EventId` I need the methods `toString()` and `toBinary()`. I could work around and always recreate the uuid object from a string to then turn it into binary, but that's some overhead.

I'm open for suggestions or PR's though.

Comment (enumag) [reply src/Util/Uuid.php]: Thanks for explanation. I think this is good to merge now, we can revisit these details later.

**Comment by enumag** [comment]
@prolic Do you know how to validate that a string is a valid GUID and not a UUID?

**Comment by prolic** [comment]
Every UUID will get accepted, the only difference is in binary encoding. So
as long as you treat only the string representation on your side (and not
the binary form) you are safe to use uuid.
Sascha-Oliver Prolic


Am Do., 22. Nov. 2018 um 18:43 Uhr schrieb Jáchym Toušek <
notifications@github.com>:

**Pull request merged**